### PR TITLE
Fix particle rk4 integration box periodicity correction

### PR DESCRIPTION
**Pull request by gassmoeller**
The correction for particle locations at periodic boundaries for the RK4 integration scheme was broken for the box geometry, possibly since its introduction, but certainly since #4264.

Cartesian geometries are special (compared to the spherical shell), because a periodicity correction for a position (= a translation) is different from a velocity (= do nothing). For a spherical shell both corrections are identical (= the same rotation). RK2 is not affected, because its temporary variable is a location (the location at the beginning of the time step), so the correction was applied correctly.

The fix I implemented is to separate the intermediate variables into locations and velocities, and let each geometry model decide how to handle the different categories. Boxes will now ignore velocities and only correct locations. Spherical shells will correct both the same way. Other corrections for other geometries are possible.

### CODE REVIEW

File: include/aspect/geometry_model/interface.h [hunk 1]
```diff
@@ -315,10 +315,15 @@ namespace aspect
...
+         * all locations given in @p connected_positions (if any where given) and all
+         * velocities given in @p connected_velocities in the same way.
+         * Adjusting both of these allows to adjust related temporary variables,
          * e.g. the intermediate results of an ordinary differential equation solver
-         * that are used to compute differences/directions between points.
+         * that are used to compute differences/directions between points. The reason
+         * positions and velocities have to be treated separately is that some geometries
+         * have to adjust them differently across a periodic boundary, e.g. a box has
+         * to translate positions but not velocities, while a spherical shell has to
+         * rotate both.
```
Comment (bangerth) [inline include/aspect/geometry_model/interface.h:326]: The movement of points can be facilitated by a rotation, but I think it's a bit confusing to say that they both have to be "rotated". Perhaps say that the points have to be moved and the velocities rotated?

File: source/geometry_model/box.cc [hunk 2]
```diff
@@ -196,7 +196,8 @@ namespace aspect
     template <int dim>
     void
     Box<dim>::adjust_positions_for_periodicity (Point<dim> &position,
-                                                const ArrayView<Point<dim>> &connected_positions) const
+                                                const ArrayView<Point<dim>> &connected_positions,
+                                                const ArrayView<Tensor<1, dim>> &/*connected_velocities*/) const
```
Comment (bangerth) [inline source/geometry_model/box.cc:200]: Nice job, by the way, using different types for positions and velocities!

File: source/geometry_model/spherical_shell.cc [hunk 3]
```diff
@@ -403,6 +404,9 @@ namespace aspect
 
           for (auto &connected_position: connected_positions)
             connected_position = rotation_matrix * connected_position;
+
+          for (auto &connected_velocity: connected_velocities)
+            connected_velocity = rotation_matrix * connected_velocity;
```
Comment (bangerth) [inline source/geometry_model/spherical_shell.cc:409]: This is interesting: The movement of points is the same operation as the rotation of the velocities. Is this always the case for arbitrary opening angles of the sector, or is this a coincidence?

File: source/particle/integrator/rk_4.cc [hunk 4]
```diff
@@ -74,107 +74,112 @@ namespace aspect
...
 
-                Point<dim> new_location = k[0] + 0.5 * k[1];
+                Point<dim> new_location = old_location + 0.5 * k[0];
 
                 // Check if we crossed a periodic boundary and if necessary adjust positions
                 if (geometry_has_periodic_boundary)
                   this->get_geometry_model().adjust_positions_for_periodicity(new_location,
-                                                                              ArrayView<Point<dim>>(&k[0],2));
+                                                                              ArrayView<Point<dim>>(&old_location,1),
+                                                                              ArrayView<Tensor<1,dim>>(&k[0],1));
```
Comment (bangerth) [inline source/particle/integrator/rk_4.cc:95]: I think you can get all of this via
```suggestion
                                                                              make_array_view(old_location),
                                                                              make_array_view(k[0]));
```

File: source/geometry_model/spherical_shell.cc (hunk 3 above)
Comment (gassmoeller) [reply source/geometry_model/spherical_shell.cc:409]: As far as I can tell by intuition it should always be the same transformation independent of opening angle. 
I suspect it is some underlying property of the coordinate transformation, but I did not write it down (likely related to the derivatives of the coordinate axes with respect to each other).

File: source/particle/integrator/rk_4.cc (hunk 4 above)
Comment (gassmoeller) [reply source/particle/integrator/rk_4.cc:95]: That seems to give me ArrayView<double>, which is not what I need for the other function. I would prefer to keep the type.

Comment (bangerth) [reply source/particle/integrator/rk_4.cc:95]: Oh, I see. Then I meant `make_array_view(k)`, which creates an array view for an array with one element of the type of the element you pass. But you are right that it does not matter.

**Comment by bangerth** [comment]
Please squash the last commit into the first and merge yourself!

**Comment by gassmoeller** [comment]
I squashed the commit, but I ll need one approving review in order to merge.

**Review by MFraters — APPROVED** [review]
The changes look reasonable to me, and since @bangerth already approved it in the comments I will approve and merge it.

**Pull request merged**